Recursive Ref Causes "String index out of range"

### Version

I'm using 4.5.11 of the vertx-openapi library, which off course transitively depends on vertx-json-schema.

### Context

I have openapi files that have $refs to separate json schema files (using draft 4 currently), which are provided to the `OpenAPIContract::fromContract` by way of the `additionalContractFiles` argument. That's kind of unrelated, but the important detail is that I have an OpenAPI file with a $ref to a schema that defines a tree structure. That tree schema for the sake of simplicity can be reduced to the following to cause an exception:

```
type: object
properties:
  children:
    type: array
      $ref: '#'
```

The exception in question is this:

```
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 0
	at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:47)
	at java.base/java.lang.String.charAt(String.java:693)
	at io.vertx.json.schema.impl.JsonRef.resolveUri(JsonRef.java:326)
	at io.vertx.json.schema.impl.JsonRef.resolve(JsonRef.java:211)
	at io.vertx.json.schema.impl.JsonRef.resolveUri(JsonRef.java:338)
	at io.vertx.json.schema.impl.JsonRef.resolve(JsonRef.java:211)
	at io.vertx.json.schema.impl.JsonRef.resolveUri(JsonRef.java:338)
	at io.vertx.json.schema.impl.JsonRef.resolve(JsonRef.java:211)
	at io.vertx.json.schema.impl.SchemaRepositoryImpl.resolve(SchemaRepositoryImpl.java:243)
	at io.vertx.openapi.contract.OpenAPIVersion.lambda$resolve$4(OpenAPIVersion.java:110)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$4(ContextImpl.java:192)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextImpl$1.execute(ContextImpl.java:221)
	at io.vertx.core.impl.WorkerTask.run(WorkerTask.java:56)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:81)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
```

### Do you have a reproducer?

I do not have a reproducer, but the problem is pretty obvious if you look at the `JsonRef::resolveUri` function. There is no isEmpty check on the path before accessing the first character.

### Steps to reproduce

Look at [this if statement](https://github.com/eclipse-vertx/vertx-json-schema/blob/81ad194af0b12567a04df4d11bcd7b0fe6eac81f/src/main/java/io/vertx/json/schema/impl/JsonRef.java#L328) and try to tell me that there's a guarantee that path will always be nonempty. If you have a `$ref: '#'`, then the parts array is going to initialized as `["#", ""]` and you'll run into the exception above at the linked if statement because the second element is empty.

The temporary solution to fix this was to change my ref to `$ref: '#/'`. It provides the same behavior as the '#' by itself, but just the pound symbol should be supported and I'm not sure why the existing tests aren't triggering the same exception that I'm running into.

### Solutions

There seems to be multiple solutions to fix this, but in the end a pound symbol by itself should just trigger `anchors.get(prefix);` in the end to retrieve the current schema. Both solutions below should result in the same behavior and both changes are to the `JsonRef::resolveUri` function.

Option 1:

We could make hashPresent set to false when the passed uri is "#".

```
    final boolean hashPresent = parts.length == 2 && parts[1] != null && !parts[1].isEmpty();
```

Option 2:

We could add a safety check before accessing the first character of the path.

```
    if (hashPresent && !path.isEmpty() && path.charAt(0) != '/') {
```

### Extra

* Anything that can be relevant such as OS version, JVM version

--- a/src/main/java/io/vertx/json/schema/impl/JsonRef.java
+++ b/src/main/java/io/vertx/json/schema/impl/JsonRef.java
@@ -320,7 +320,9 @@ public final class JsonRef {
     //  [prefix, path]
     final String[] parts = uri.split("#", 2);
 
-    final boolean hashPresent = parts.length == 2 && parts[1] != null;
+    // an empty fragment ("#" or "id#") is equivalent to no fragment at all: the uri refers
+    // to the schema identified by the prefix, e.g.: the root schema when the prefix is empty
+    final boolean hashPresent = parts.length == 2 && parts[1] != null && !parts[1].isEmpty();
 
     final String prefix = parts[0];
     final String path = hashPresent ? parts[1] : null;

--- a/src/test/java/io/vertx/tests/impl/RefTest.java
+++ b/src/test/java/io/vertx/tests/impl/RefTest.java
@@ -204,6 +204,83 @@ class RefTest {
   }
 
   @Test
+  public void testRefToRoot() {
+    JsonObject schema = new JsonObject()
+      .put("type", "object")
+      .put("properties", new JsonObject()
+        .put("children", new JsonObject()
+          .put("type", "array")
+          .put("items", new JsonObject()
+            .put("$ref", "#"))));
+
+    JsonObject resolved = JsonRef.resolve(schema);
+
+    JsonObject items = resolved
+      .getJsonObject("properties")
+      .getJsonObject("children")
+      .getJsonObject("items");
+    assertEquals("object", items.getString("type"));
+    // the ref points back to the root schema itself
+    assertNotNull(
+      items
+        .getJsonObject("properties")
+        .getJsonObject("children")
+        .getJsonObject("items"));
+  }
+
+  @Test
+  public void testRefToRootWithId() {
+    JsonObject schema = new JsonObject()
+      .put("$id", "http://www.example.com/tree/")
+      .put("type", "object")
+      .put("properties", new JsonObject()
+        .put("children", new JsonObject()
+          .put("type", "array")
+          .put("items", new JsonObject()
+            .put("$ref", "#"))));
+
+    JsonObject resolved = JsonRef.resolve(schema);
+
+    JsonObject items = resolved
+      .getJsonObject("properties")
+      .getJsonObject("children")
+      .getJsonObject("items");
+    assertEquals("object", items.getString("type"));
+    assertNotNull(
+      items
+        .getJsonObject("properties")
+        .getJsonObject("children")
+        .getJsonObject("items"));
+  }
+
+  @Test
+  public void testRefToRootThroughRepository() {
+    SchemaRepository repo =
+      SchemaRepository.create(new JsonSchemaOptions().setBaseUri("http://vertx.io").setDraft(Draft.DRAFT4));
+    JsonObject schema = new JsonObject()
+      .put("type", "object")
+      .put("properties", new JsonObject()
+        .put("children", new JsonObject()
+          .put("type", "array")
+          .put("items", new JsonObject()
+            .put("$ref", "#"))));
+    repo.dereference(JsonSchema.of(schema.copy()));
+
+    JsonObject resolved = repo.resolve(schema);
+
+    JsonObject items = resolved
+      .getJsonObject("properties")
+      .getJsonObject("children")
+      .getJsonObject("items");
+    assertEquals("object", items.getString("type"));
+    assertNotNull(
+      items
+        .getJsonObject("properties")
+        .getJsonObject("children")
+        .getJsonObject("items"));
+  }
+
+  @Test
   void testSerialization() {
     SchemaRepository repo =
       SchemaRepository.create(new JsonSchemaOptions().setBaseUri("http://vertx.io").setDraft(Draft.DRAFT202012));

